yt/utilities/lib/cyoctree.pyx

```diff
@@ -1,679 +1,756 @@
-            leftx = node.left_edge[0]
-            lefty = node.left_edge[1]
-            leftz = node.left_edge[2]
-            rightx = node.right_edge[0]
-            righty = node.right_edge[1]
-            rightz = node.right_edge[2]
-
-            # We have to generate cell locations
-            dx = (rightx - leftx) / self._num_cells_per_dim
-            dy = (righty - lefty) / self._num_cells_per_dim
-            dz = (rightz - leftz) / self._num_cells_per_dim
-
-            # Loop through each cell and calculate the contribution, l is the
-            # number of the cell we are in
-            l = 0
-            for i in range(self._num_cells_per_dim):
-                for j in range(self._num_cells_per_dim):
-                    for k in range(self._num_cells_per_dim):
-                        diff_x = (leftx + (i + 0.5) * dx - posx)
-                        diff_x *= diff_x
-                        diff_y = (lefty + (j + 0.5) * dy - posy)
-                        diff_y *= diff_y
-                        diff_z = (leftz + (k + 0.5) * dz - posz)
-                        diff_z *= diff_z
-
-                        voxel_hsml2 = hsml*hsml
-                        q_ij = math.sqrt((diff_x + diff_y + diff_z) /
-                                                                    voxel_hsml2)
-
-                        if use_normalization:
-                            buff_den[node.leaf_id + l] += (prefactor_norm *
-                                                           self.kernel(q_ij))
-                        buff[node.leaf_id + l] += prefactor * self.kernel(q_ij)
-                        l += 1
-
-    def interpolate_sph_cells(self, np.float64_t[:] buff,
-                              np.float64_t[:] buff_den, np.float64_t[:] posx,
-                              np.float64_t[:] posy, np.float64_t[:] posz,
-                              np.float64_t[:] pmass, np.float64_t[:] pdens,
-                              np.float64_t[:] hsml, np.float64_t[:] field,
-                              kernel_name="cubic", int use_normalization=0):
+            for i in range(8):
+                child_node = tree.children[8*num_node + i]
+
+                diff_x = tree.node_positions[3*child_node] - posx
+                diff_y = tree.node_positions[3*child_node+1] - posy
+                diff_z = tree.node_positions[3*child_node+2] - posz
+                diff = math.sqrt(diff_x*diff_x + diff_y*diff_y + diff_z*diff_z)
+
+                sx = tree.size[0] / (2**tree.depth[child_node])
+                sy = tree.size[1] / (2**tree.depth[child_node])
+                sz = tree.size[2] / (2**tree.depth[child_node])
+                size = math.sqrt(sx*sx + sy*sy + sz*sz)
+
+                if diff - size - hsml < 0:
+                    self.smooth_onto_cells(buff, buff_den, posx, posy, posz,
+                                    hsml, prefactor, prefactor_norm,
+                                    child_node, use_normalization=use_normalization)
```

Comment:
Note: you could spare some CPU cycles by computing 
```python
local_dx = tree.size[0] / 2.**tree.depth[root_node]
```
when hitting the root node, and then pass `local_dx/2` as an argument to `smooth_onto_cells`.
This could look like
```suggestion
                # THIS CODE IS NOT COMPLETE, `local_d{xyz}` should have been computed at the root node.
                diff_x = tree.node_positions[3*child_node] - posx
                diff_y = tree.node_positions[3*child_node+1] - posy
                diff_z = tree.node_positions[3*child_node+2] - posz
                diff = math.sqrt(diff_x*diff_x + diff_y*diff_y + diff_z*diff_z)

                sx = local_dx
                sy = local_dy
                sz = local_dz
                size = math.sqrt(sx*sx + sy*sy + sz*sz)

                if diff - size - hsml < 0:
                    self.smooth_onto_cells(buff, buff_den, posx, posy, posz,
                                    hsml, prefactor, prefactor_norm,
                                    child_node, use_normalization=use_normalization,
                                    local_dx=local_dx/2, local_dy=local_dy/2, local_dz=local_dz/2)
```